econml/metalearners.py

```diff
@@ -416,164 +413,23 @@ def effect(self, X):
 
         Returns
         -------
-        τ_hat : array-like, shape (m, )
-            Matrix of heterogeneous treatment effects for each sample.
+        τ_hat : matrix, shape (m, d_y, n_t-1), n_t is number of treatments
+            Constant marginal CATE of each treatment on each outcome for each sample X[i].
+            Note that when Y is a vector rather than a 2-dimensional array,
+            the corresponding singleton dimensions in the output will be collapsed
         """
-        # Check inputs
         X = check_array(X)
-        tau_hat = self.overall_model.predict(X)
-        return tau_hat
-
-    def marginal_effect(self, X):
-        """Calculate the heterogeneous marginal treatment effect.
-
-        For binary treatments, it returns the same as `effect`.
-
-        Parameters
-        ----------
-        X : matrix, shape (m × dₓ)
-            Matrix of features for each sample.
-
-        Returns
-        -------
-        τ_hat : array-like, shape (m, )
-            Matrix of heterogeneous treatment effects for each sample.
-        """
-        return self.effect(X)
+        taus = []
+        for model in self.final_models:
+            taus.append(model.predict(X))
+        taus = np.column_stack(taus).reshape((-1, self._d_t - 1,) + self._d_y)  # shape as of m*d_t*d_y
+        if self._d_y:
+            taus = transpose(taus, (0, 2, 1))  # shape as of m*d_y*d_t
+        return taus
```

Comment:
Looks like very similar logic to this shows up in a few places.  Would it be worthwhile to create a common base class so that the logic doesn't need to be repeated?